Sources/Basic/OutputByteStream.swift

```diff
@@ -55,27 +60,86 @@ public class OutputByteStream: TextOutputStream {
         return buffer.count
     }
 
-    /// The contents of the output stream.
-    ///
-    /// This method implicitly flushes the stream.
-    public final var bytes: ByteString {
-        flush()
-        return ByteString(self.buffer)
+    /// Currently available buffer size.
+    private var availableBufferSize: Int {
+        return buffer.capacity - buffer.count
     }
-    
+
+     /// Clears the buffer maintaining current capacity.
+    private func clearBuffer() {
+        buffer.removeAll(keepingCapacity: true)
+    }
+
     // MARK: Data Output API
 
     public final func flush() {
+        writeImpl(buffer)
+        clearBuffer()
+        flushImpl()
+    }
+
+    func flushImpl() {
         // Do nothing.
     }
 
+    func writeImpl<C: Collection where C.Iterator.Element == UInt8>(_ bytes: C) {
```

Comment:
Any call to writeImpl would need to be devirtualized first. Only then could it call into a specialized implementation.

There is an @_specialize attribute though, which you could use to specialize a particular subclass writeImpl for a particular Collection type.

I would go with the light-weight UnsafeBufferPointer if you can.